OPHD/UI/IconGrid.h

```diff
@@ -108,6 +109,7 @@ class IconGrid : public Control
 	NAS2D::Vector<int> mGridSize; /**< Dimensions of the grid that can be contained in the IconGrid with the current Icon Size and Icon Margin. */
 
 	IconItemList mIconItemList; /**< List of items. */
+	std::map<std::string, std::vector<std::tuple<std::string, int>>> mResourceCosts; /**< List of resources required to build the item. */
```

Comment:
This is an interesting and slightly surprising data structure to accomplish the given task. I can see how this would work, and be fairly general. It also seems a bit heavy in some ways.

The `std::vector` portion contains multiple `std::string`, and those's strings are likely to be identical for every `std::vector` instance, yet there is no sharing. Each `std::string` is a separate copy taking up additional memory.

On the other hand, by storing the `std::string` and the `int` separately means we need to dynamically generate concatenated strings during both sizing and display. If the full string was generated up front, that would eliminate the need for dynamic creation during display. Additionally, the string sizes could potentially be calculated up front as well, so they wouldn't need to be recalculated every frame.

As a further minor note, although there would be just as many strings as before, the combination of the resource name and cost are more likely to be unique, so in that sense there would be less duplication. There's no savings here, rather, there's more justification for the current consumption.

One downside to generating the strings early, is if you wanted to format the resource name differently from the resource cost, it would then be harder to do. Perhaps the names and cost would be in a different color, or font size, or have different alignment.

----

As for why I found the solution surprising, I think I expected the entries of `IconItemList` to contain the extra data, rather than a separate lookup structure. That would simplify the looping, since all data would be stored together in the same structure, so there would be no need for an additional lookup, and no need for error checking against a failed lookup.

Perhaps the other surprising bit was the custom data structure using only standard library types. I was perhaps expecting `StorableResources` and `PopulationRequirements` as the types for additional fields. This idea of course being more natural if they are added to `IconItemList`, rather than the `std::map` structure.

----

Even if the data is stored as separate components (and dynamically generated during display, rather than storing a pre-generated `std::string`), the sizing can still be done up front. The needed size would be known as soon as the `IconItemList` field is set/updated. Here I'm kind of assuming it would be set up front, and not really modified after. I haven't really checked the source for that point. It's probably not actually written with that in mind.

Actually, I'm wondering about the design having an `addItem` function. Maybe there should only be the option to construct `IconGrid` with all items at once. It's harder to optimize these things when the data is mutable, and in many cases, mutability isn't actually needed. The game knows early on what all the options and costs are, with changes not being common.